File: services/payments-worker/tracing.py
"""OpenTelemetry setup + W3C trace-context propagation helpers for the worker."""

import os

from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.propagate import extract, inject
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.trace import format_trace_id


def init_tracer() -> None:
    """Wire an OTLP/gRPC exporter from OTEL_EXPORTER_OTLP_ENDPOINT (no-op if unset)."""
    endpoint = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT")
    if not endpoint:
        return
    name = os.getenv("OTEL_SERVICE_NAME", "payments-worker")
    provider = TracerProvider(resource=Resource.create({"service.name": name}))
    provider.add_span_processor(
        BatchSpanProcessor(OTLPSpanExporter(endpoint=endpoint, insecure=True))
    )
    trace.set_tracer_provider(provider)


def extract_context(fields: dict):
    """Continue the distributed trace carried in a stream message's fields."""
    carrier = {}
    if "traceparent" in fields:
        carrier["traceparent"] = fields["traceparent"]
    return extract(carrier)


def inject_context() -> dict:
    """Serialize the current trace context for embedding in the next message."""
    carrier: dict = {}
    inject(carrier)
    return carrier


def current_trace_id() -> str:
    ctx = trace.get_current_span().get_span_context()
    if ctx.trace_id:
        return format_trace_id(ctx.trace_id)
    return ""

File: services/orders-api/tracing.go
package main

import (
	"context"
	"log"
	"os"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/propagation"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
	"go.opentelemetry.io/otel/trace"
)

// initTracer wires an OTLP/gRPC trace exporter from OTEL_EXPORTER_OTLP_ENDPOINT.
// Returns a no-op shutdown when tracing is unconfigured (e.g. local tests).
func initTracer(ctx context.Context) func() {
	endpoint := os.Getenv("OTEL_EXPORTER_OTLP_ENDPOINT")
	if endpoint == "" {
		return func() {}
	}
	exp, err := otlptracegrpc.New(ctx,
		otlptracegrpc.WithEndpoint(endpoint),
		otlptracegrpc.WithInsecure(),
	)
	if err != nil {
		log.Printf("otel exporter: %v", err)
		return func() {}
	}
	name := os.Getenv("OTEL_SERVICE_NAME")
	if name == "" {
		name = "orders-api"
	}
	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exp),
		sdktrace.WithResource(resource.NewWithAttributes(semconv.SchemaURL, semconv.ServiceName(name))),
	)
	otel.SetTracerProvider(tp)
	otel.SetTextMapPropagator(propagation.TraceContext{})
	return func() { _ = tp.Shutdown(context.Background()) }
}

// injectTrace serializes the active trace context into a carrier map suitable
// for embedding in a Redis stream message (so consumers can continue the trace).
func injectTrace(ctx context.Context) map[string]string {
	carrier := propagation.MapCarrier{}
	otel.GetTextMapPropagator().Inject(ctx, carrier)
	return carrier
}

// traceID returns the current span's trace id (hex) or "" if none.
func traceID(ctx context.Context) string {
	sc := trace.SpanContextFromContext(ctx)
	if sc.HasTraceID() {
		return sc.TraceID().String()
	}
	return ""
}
